### DISCUSSION TITLE
How to fix delta treats everything as change when using it as a `diff` alternative?

### DISCUSSION THREAD

**Discussion by wytsai7660** [dandavison/delta#1555] [Q&A] [answered]
Hello! I'm using MacOS, and I downloaded `delta` via homebrew.

When testing this tool with the command:

```shell
delta test.c test2.c
``` 

I get the output:

<img width="922" alt="Screenshot 2023-11-02 at 4 38 51 AM" src="https://github.com/dandavison/delta/assets/116703995/87eb305e-196a-4253-91f5-55af6fef8939">

I'm wondering why everything is being treated as change.

Additionally, I noticed that the command failed for some reason (as indicated by the color of the prompt characters in the powerlevel10k theme), and I also have no idea why.

**Comment by dandavison** [comment]
Hi @wytsai7660,

It's failing (exiting with status code 1) because it is emulating the [behavior of diff](https://stackoverflow.com/questions/6971284/what-are-the-error-exit-values-for-diff).

For your first question, can you post the raw contents of the files please?

**Reply by wytsai7660** [reply]
Oh, I got it. Thanks for the explanation!

For my first question, here are the contents of `test.c` and `test2.c`

`test.c`:
```c
#include <stdio.h>

int main() {
  int a = 1, b = 2, c;
  c = a + b;
  printf("c = %d\n", c);
}
// with an extra new line here
```

`test2.c`:
```c
#include <stdio.h>
#include <stdlib.h>

int main() {
  int a = 1, b = 4, c;
  c = 2 * b + a;
  printf("c = %d\n", c);
  return 0;
}
// with an extra new line here
```

**Reply by wytsai7660** [reply]
@dandavison  Hello, do you have any updates about this issue?

**Reply by dandavison** [reply]
I don't seem to be able to reproduce that:

<img width="299" alt="image" src="https://github.com/dandavison/delta/assets/52205/2d660683-b84d-449a-b4aa-3855c7a909a4">

**Reply by wytsai7660** [reply]
I apologize for the delayed response. I was occupied with work lately and inadvertently overlooked our discussion.

The problem persists even after reinstalling `delta`. I've ensured that there is only my user information in the .gitconfig file. Are there any potential reasons for this issue?

**Reply by dandavison** [reply]
Would you mind following exactly the same steps as I used here? I.e. `delta --no-gitconfig a b`

```
/tmp cat > a
#include <stdio.h>

int main() {
  int a = 1, b = 2, c;
  c = a + b;
  printf("c = %d\n", c);
}
// with an extra new line here
/tmp cat > b
#include <stdio.h>
#include <stdlib.h>

int main() {
  int a = 1, b = 4, c;
  c = 2 * b + a;
  printf("c = %d\n", c);
  return 0;
}
// with an extra new line here
/tmp delta --no-gitconfig a b
/tmp delta --version
delta 0.16.5
```

<img width="306" alt="image" src="https://github.com/dandavison/delta/assets/52205/65c03ec4-2db1-4d92-9972-95ea91041abe">

**Comment by wytsai7660** [comment]
It worked when I create and test with the new files! even without the --no-gitconfig flag.

<img width="938" alt="Screenshot 2023-12-23 at 5 57 56 AM" src="https://github.com/dandavison/delta/assets/116703995/83f3ca0a-178c-4792-bc70-4967df0f1138">

Based on my test, it seems that it only doesn't work on those two specific files (even when they are moved to another folder, renamed, or have their file extension removed). Are you willing to investigate further into those two files? How should I share them with you?

**Reply by dandavison** [reply]
Can you paste the output of `cat -A` or `bat -A` on the files?

**Reply by wytsai7660** [reply]
Sure, here they are

<img width="938" alt="Screenshot 2023-12-23 at 6 59 04 AM" src="https://github.com/dandavison/delta/assets/116703995/1840a913-7463-4cf1-b5f5-23a11742e25a">

<img width="938" alt="Screenshot 2023-12-23 at 6 59 23 AM" src="https://github.com/dandavison/delta/assets/116703995/10233aa2-528b-467a-b5f0-880d912f60b5">

**Reply by dandavison** [reply, marked as answer]
OK, so it looks like one of the files has Windows-style line endings (CR LF) and the other one has Unix/Mac-style (LF).

**Reply by wytsai7660** [reply]
Oh, I see. That makes perfect sense. So when we can't find the difference, maybe it was caused by some kind of control character (not visible).
This answered my confusion. Thank you so much.

On the other hand, maybe we can open an issue of a feature request: When such cases happen, output the control character (e.g., ␍). How do you think?

**Reply by dandavison** [reply]
I think you're right that delta should do better here. It might fall under this open issue: https://github.com/dandavison/delta/issues/522

**Reply by wytsai7660** [reply]
OK, thank you again for proactively responding to my question. Appreciate that.
